First pull request of the module

GianniLunardi opened the pull request:
First pull request after the refractoring of Safe MPC in [VBOC](https://github.com/GianniLunardi/VBOC/tree/icra).
Properties:

- all controllers are implemented
- only one model (triple pendulum)
- tests on naive and htwa

GianniLunardi committed:
[add] ignore elements

GianniLunardi committed:
[add] package initialization

GianniLunardi committed:
[add] first test on the naive mpc controller (wip)

GianniLunardi committed:
[minor] ignore pycache folders

GianniLunardi committed:
moved or deleted scripts

GianniLunardi committed:
safe-mpc package:

- init
- abstract model and ocp
- dynamics simulator

GianniLunardi committed:
[add] of the common definitions (like ROOT dir)

GianniLunardi committed:
[add] double and triple pendulum models

GianniLunardi committed:
[add] some tests to be compared with the results in VBOC repo (some changes are required)

GianniLunardi committed:
[add] main controller (Naive, ST, S/HTWA, Receding)

GianniLunardi committed:
[add] abort controllers

GianniLunardi committed:
ignore all generated files from acados/casadi and others

GianniLunardi committed:
fix on dimensions

GianniLunardi committed:
fixes:

- return copies of arrays/values
- running + terminal constraints
- NN class
- set NN model

GianniLunardi committed:
improvements on the class inheritance

GianniLunardi committed:
del

GianniLunardi committed:
amodel to improve clarity

GianniLunardi committed:
test files to check naive controllers works and dynamics check (TO BE CLEANED)

GianniLunardi committed:
[add] debug class

GianniLunardi committed:
[add] nn model for triple pendulum

GianniLunardi committed:
[add] params config

GianniLunardi committed:
[add] parser for the config file

GianniLunardi committed:
[add] test that all the objects 'controllers' of the safe mpc can be created

GianniLunardi committed:
[add] simulator parameters

GianniLunardi committed:
fix the nested relationships model -> simulator -> controller

GianniLunardi committed:
update readme

GianniLunardi committed:
[minor] fix on the horizon (for the htwa)

GianniLunardi committed:
[minor] remove the x_sim from the plot

GianniLunardi committed:
[add] test for the htwa controller

GianniLunardi committed:
[fix] re-initialize u_guess to zero

until now, if one initialization failed, then u_guess is set to nan and all the next cases results in failures

GianniLunardi committed:
[fix] velocity problem for the NN model

GianniLunardi committed:
[moved] to relu folder

GianniLunardi committed:
[add] triple pendulum viability kernel with elu activation function

GianniLunardi committed:
adaptable x_sim length and grid activated by default

GianniLunardi committed:
[minor] change parameter names

GianniLunardi committed:
[minor]

GianniLunardi committed:
add different tests for the viable initial guess

- htwa
- st with hardening soft weight
- st with increasing soft weight

GianniLunardi committed:
minor changes:

- l4c
- SX to MX symbols
- L2 to L1 norm for soft constraint

GianniLunardi committed:
minor changes:

- set the initial viable state
- L2 to L1 norm for soft constraint
- check dynamics constraint

andreadelprete commented on data/nn_model/mean_3dof_vboc:
I would avoid commiting the NN weights in the git repo

GianniLunardi replied:
Okay, in this case we need to find a turn around to provide the NN weights, since are necessary for run the scripts

andreadelprete replied:
You could just have them stored locally on your machine and use a GDrive folder for backup.

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+        pass
+
+    def checkStateConstraints(self, x):
+        return np.all((x >= self.x_min) & (x <= self.x_max))
```
Shouldn't you use "and" instead of "&"? Using "&" results in an eager evaluation of the two operands, which is just a waste of time in this context.

andreadelprete replied:
Or even better, you should use the np.logical_and function. https://numpy.org/doc/stable/reference/generated/numpy.logical_and.html

GianniLunardi replied:
`np.logical_and` works

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+        return np.all((x >= self.x_min) & (x <= self.x_max))
+
+    def checkControlConstraints(self, u):
+        return np.all((u >= self.u_min) & (u <= self.u_max))
```
Shouldn't you use "and" instead of "&"? Using "&" results in an eager evaluation of the two operands, which is just a waste of time in this context.

andreadelprete replied:
Or even better, you should use the np.logical_and function. https://numpy.org/doc/stable/reference/generated/numpy.logical_and.html

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+        return self.checkStateConstraints(x) and self.checkControlConstraints(u)
+
+    def checkSafeConstraints(self, x):
+        return True if float(self.nn_func(x, self.params.alpha).toarray()) >= 0. else False
```
This is equivalent to: 
```
        return float(self.nn_func(x, self.params.alpha).toarray()) >= 0. 
```

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+        model = NeuralNetDIR(self.nx, (self.nx - 1) * 100, 1).to(device)
+        model.load_state_dict(torch.load(self.params.NN_DIR + 'model_3dof_vboc', map_location=device))
+        mean = torch.load(self.params.NN_DIR + 'mean_3dof_vboc')
+        std = torch.load(self.params.NN_DIR + 'std_3dof_vboc')
```
Having hard-coded names that refer to the 3dof model does not seem appropriate in this class

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+            x_sim[i + 1] = self.simulate(x_sim[i], u[i])
+        # Check if the rollout state trajectory is almost equal to the optimal one
+        print('Norm diff: ', np.linalg.norm(x - x_sim))
+        return True if np.linalg.norm(x - x_sim) < 1e-3 else False
```
This should be `return (np.linalg.norm(x - x_sim) < 1e-3)`

andreadelprete replied:
And instead of being hard-coded, the threshold 1e-3 should be inside a configuration file

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -0,0 +1,322 @@
+        # self.ocp.solver_options.hessian_approx = "EXACT"
+        self.ocp.solver_options.nlp_solver_max_iter = 200
+        self.ocp.solver_options.qp_solver_iter_max = self.params.qp_max_iter
+        self.ocp.solver_options.globalization = "MERIT_BACKTRACKING"
```
All the values of the parameters should be moved to a configuration file

andreadelprete commented on src/safe_mpc/controller.py:
```diff
@@ -0,0 +1,121 @@
+
+    def step(self, x):
+        status = self.solve(x)
+        if status == 0 and self.model.checkRunningConstraints(self.x_temp, self.u_temp) and \
```
self.x_temp should contain the whole state trajectory, but the method checkRunningConstraints seems to be checking a single state-control pair. How can this work?

andreadelprete replied:
BTW: I think you should probably check the whole trajectory here

GianniLunardi replied:
`checkRunningConstraints` checks on all the trajectory, for a single state-control pair we need to provide only that pair as input of the method

andreadelprete commented on src/safe_mpc/controller.py:
```diff
@@ -0,0 +1,121 @@
+        # Terminal constraint
+        self.ocp.ocp_solver.cost_set(self.N, "Zl", 1e5 * np.ones((1,)))
+        # Receding constraint
+        self.ocp.ocp_solver.cost_set(self.r, "Zl", 1e8 * np.ones((1,)))
```
These weights (1e5 and 1e8) should be moved to a configuration file

andreadelprete commented on src/safe_mpc/controller.py:
```diff
@@ -0,0 +1,121 @@
+class RecedingController(NaiveController):
+    def __init__(self, simulator):
+        super().__init__(simulator)
+        self.x_viable = None                # TODO: to be initialized
```
Not clear what the role of `x_viable` is. Maybe add a comment to explain?

GianniLunardi replied:
I will add a comment on this. The idea is to store in this variable the last state of the predicted trajectory, after checking if it is viable. In this way the MPC loop can be interrupted in any time, but we have in memory the state from which we trigger the safe abort

andreadelprete commented on src/safe_mpc/abort.py:
```diff
@@ -0,0 +1,73 @@
+from .abstract import AbstractController
+
+
+class SafeBackupController(AbstractController):
```
I don't know if it makes sense that the backup controllers are child classes of `AbstractController`, because they do a very different thing. The `AbstractController` is an MPC, while the backup controllers rely on solving OCP's up to convergence. What do you think?

GianniLunardi replied:
From my point of view, `AbstractController` can refer both to single OCP or MPC, since the initialization of the Acados OCP object is almost equal. To solve an OCP, one can set the initial guess and call the method `solve`, while for the MPC cycle we call `step`, which is defined differently for the controllers. In this way, we can avoid duplicated code fragments.

andreadelprete replied:
ok

andreadelprete reviewed (changes requested):
Hi @GianniLunardi , sorry it took me a while to review this. Anyway, I think the code is much cleaner now. I still have a few concerns though, so please try to answer my questions below and address the associated comments.

GianniLunardi committed:
[deleted] old test files

GianniLunardi committed:
start ignoring nn models

GianniLunardi committed:
[minor] some changes and other parameters

GianniLunardi committed:
from SX to MX

GianniLunardi committed:
changes

- return a flag in the initialization method
- check safe constraint on the initial guess on the WA methods
- get last viable state
- minor correction on RecedingConstraint

GianniLunardi committed:
convergence objective and stats (profiling)

GianniLunardi committed:
different tests:

- abort task
- constrained (STWA, HTWA, Receding)
- initial guess of constrained problems
- refiniment of the initial guess (WIP)

GianniLunardi committed:
ignore npy and nn model

GianniLunardi committed:
code improvements

GianniLunardi committed:
compute the 99 quantile (maybe should be changes into max)

GianniLunardi committed:
[add] nn_models folder

GianniLunardi committed:
[deleted] old params file

GianniLunardi committed:
add yaml files and update the parser

GianniLunardi committed:
provide minor improvements

GianniLunardi committed:
[add] the main script

it will be used to run any controller/test given the arguments

GianniLunardi committed:
[WIP] add main script to run any controller

GianniLunardi committed:
[minor] change consistency with new parser

GianniLunardi committed:
[add] nn_models folder

GianniLunardi committed:
[add] new scripts:

- gravity compensation through minimization problem
- test VBOC-like abort
- initialization using IPOPT

GianniLunardi committed:
minor changes

GianniLunardi committed:
[add] ipopt controller

GianniLunardi committed:
change the VBOC Controller

GianniLunardi committed:
[wip] test urdf pinocchio models against handwritten dynamics

GianniLunardi committed:
add options for VBOC-like controller (will be removed)

GianniLunardi committed:
add terminal constraint option for ipopt controller (to slow)

GianniLunardi committed:
comment parameters similar to double pendulum urdf (example robot data)

GianniLunardi committed:
add better initialization

GianniLunardi committed:
Merge branch 'personal/reordering' into personal/convergence

# Conflicts:
#	.gitignore
#	config/params.yaml
#	src/safe_mpc/abstract.py
#	src/safe_mpc/controller.py
#	src/safe_mpc/parser.py
#	tests/initialization.py
#	tests/test_naive.py

GianniLunardi committed:
[remove] VBOC-like and [move] Safe-Backup to controller.py

GianniLunardi committed:
[delete] old test scripts

GianniLunardi committed:
[add] state selection tolerance

GianniLunardi committed:
minor fix (adaptation to Parameters class API)

GianniLunardi committed:
[add] safe abort to controller.py

GianniLunardi committed:
[add] argument parser for the main script

GianniLunardi committed:
minor changes:

- define number of shooting nodes inside the controller class initialization
- time fields
- viable state for all the controller (ignored for naive, st, abort)
- RuntimeError for Ipopt (to be checked)

GianniLunardi committed:
minor changes:

- moved argument parser to parser.py
- fix in the initial guess and simulate methods
- general convergence criteria for task completion
- names and folders for saved data
- abort task

GianniLunardi committed:
specify zip extension for NN parameters

GianniLunardi committed:
update installation and usage instructions

GianniLunardi committed:
list of python packages required

GianniLunardi commented:
Hi @andreadelprete,
I have addressed all the comments and reported some changes with the last 30 commits.
Main changes:

- single main script to run the tests
- deletion of the VBOC-like controller

andreadelprete commented on scripts/main.py:
```diff
@@ -0,0 +1,180 @@
+    # If ICs is active, compute the initial conditions for all the controller
+    if args['init_conditions']:
+        from scipy.stats import qmc
+        test_num = 500                      # Higher than the one reported in the configuration file
```
Shouldn't this parameter go in a configuration file?

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -56,50 +57,58 @@ def __init__(self, params):
         x_cp = deepcopy(self.x)
-        x_cp[self.nq] += 1e-8
-        vel_norm = fmax(norm_2(x_cp[self.nq:]), 1e-3)
+        x_cp[self.nq] += 1e-6
```
Replace this hard-coded value with a parameter

andreadelprete commented on src/safe_mpc/abstract.py:
```diff
@@ -131,8 +141,7 @@ def checkDynamicsConstraints(self, x, u):
         # Check if the rollout state trajectory is almost equal to the optimal one
-        print('Norm diff: ', np.linalg.norm(x - x_sim))
-        return True if np.linalg.norm(x - x_sim) < 1e-3 else False
+        return np.linalg.norm(x - x_sim) < 1e-5 * np.sqrt(n+1) 
```
Replace this hard-coded value with a parameter

andreadelprete commented on scripts/main.py:
```diff
@@ -0,0 +1,180 @@
+    simulator = SimDynamics(model)
+    controller = getattr(controllers, available_controllers[args['controller']])(simulator)
+
+    x_ref = np.array([conf.q_max - 0.05, np.pi, np.pi, 0, 0, 0])
```
I would move `x_ref` to the config file

andreadelprete commented on scripts/main.py:
```diff
@@ -0,0 +1,180 @@
+            controller.ocp_solver.cost_set(i, "zl", conf.ws_r * np.ones((1,)))
+        controller.ocp_solver.cost_set(controller.N, "zl", conf.ws_t * np.ones((1,)))
+
+        overall_success, x_guess_vec, u_guess_vec = init_guess(test_num)
```
I don't get why you try to solve 500 OCP's if then you need only `conf.test_num` initial conditions. Couldn't you keep track of how many problems you have solved and stop once you've reached `conf.test_num`?

andreadelprete reviewed (commented):
Excellent work. I've left a few minor comments below.

GianniLunardi committed:
[add] tolerance to config file

GianniLunardi committed:
ignore pdf files (e.g. images/plots)

GianniLunardi committed:
[fix] on Naive controller

the MPC shall continue even if we have roll out all the guess for the naive/st controllers

GianniLunardi committed:
save MPC results in pickles

GianniLunardi committed:
[add] computations for the mean cost increments

GianniLunardi committed:
[add] plot utils with its configuration file

GianniLunardi committed:
[add] joint target in the configuration file and model

GianniLunardi committed:
changes on the ICs computations

loops on random initial configurations until it find test_num solutions
